### PR TITLE
[kdump] Create AzureKDump class

### PR DESCRIPTION
**Pull request by aadhar-agarwal**
This change collects kdump information for Azure Linux.

With this change, we will check the 'path' variable in /etc/kdump.conf to check where information is being dumped.

If get_vm_core is set to true, collect the latest vm core created
in the last 24 hours that is <= 2GB

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [ ] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [ ] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

### CODE REVIEW

File: sos/report/plugins/kdump.py
Comment (jcastill) [inline sos/report/plugins/kdump.py:140]: Does this mean that azure kexec/kdump can only dump the vmcores into already mounted filesystems?

File: sos/report/plugins/kdump.py
Comment (jcastill) [inline sos/report/plugins/kdump.py:160]: This could potentially capture the vmcore themselves, not just the dmesg files. Capturing everything in that directory could create huge sos reports, and I'm not sure what's the benefit to capture old vmcores.

Comment (aadhar-agarwal) [reply sos/report/plugins/kdump.py]: We do want to capture the vmcore files. However, it could create really huge sos reports. I took a look at it again and created a `get_vm_core` plugin opt that is set to false by default. If this value is set to true, then we will collect the vmcore only if its size is <= 2GB and has been created in the last 24 hours. What do you think about this approach? :)

File: sos/report/plugins/kdump.py
Comment (aadhar-agarwal) [reply sos/report/plugins/kdump.py:145]: It dumps the vmcores into the rootfs by default; we wanted to start with this approach for now (by only reading the path variable in the /etc/kdump.conf file)

File: sos/report/plugins/kdump.py
Comment (jcastill) [reply sos/report/plugins/kdump.py]: I like it. In support I remember having the issue of having to ask for both sos report and to upload the vmcores. If we can do this  in one step, we can have all the data we need really easily.

File: sos/report/plugins/kdump.py
Comment (TurboTurtle) [inline sos/report/plugins/kdump.py:170]: The use of `sizelimit` here may be undesirable. With `add_copy_spec()`, this parameter tells sos to copy the last X MB of the file if it is over the given size, which would result in an unusable binary blob in the archive taking up 2GB of space.

What we would want here is to either capture a usable vmcore, or nothing at all, so rather than the use of `sizelimit`, we should get a list of potential vmcore(s) to copy, check their size and age (e.g. `os.stat()`), and only then pass the path to an `add_copy_spec()` call.

`maxage` makes sense here.

File: sos/report/plugins/kdump.py
Comment (TurboTurtle) [inline sos/report/plugins/kdump.py:134]: @pmoravec @arif-ali did we ever come up with a decision on dashes-vs-underscores for plugin options? We have a small mix today, whereas our top-level options are all dashes. I remember bringing this up on IRC a long time ago, but can't recall if we ever made a choice?

File: sos/report/plugins/kdump.py
Comment (aadhar-agarwal) [reply sos/report/plugins/kdump.py:170]: This is a great point. When I was testing actually, I saw the "File is over size limit and is bianry" log - So, I think the last X MB is not copied?

https://github.com/sosreport/sos/blob/main/sos/report/plugins/__init__.py#L1854C1-L1860C37

Please correct me if I am wrong. Based on this, I think the whole binary is copied or nothing at all is copied :)

Comment (TurboTurtle) [reply sos/report/plugins/kdump.py:170]: Oh, duh. We did change that default a while back. I swear some days I don't know where my head is at.

Comment (aadhar-agarwal) [reply sos/report/plugins/kdump.py:170]: No worries at all :) 

Glad we're on the same page!

File: sos/report/plugins/kdump.py
Comment (aadhar-agarwal) [reply sos/report/plugins/kdump.py]: Agreed!

File: sos/report/plugins/kdump.py
Comment (arif-ali) [reply sos/report/plugins/kdump.py:134]: I don't recall on any discussions, maybe it was before I got active on the project

Comment (TurboTurtle) [reply sos/report/plugins/kdump.py:134]: Yeah, the main reason is consistency for user experience. If we didn't come up with anything back then, we can start a new discussion around it and not hold up this PR.

**Review by TurboTurtle — APPROVED** [review]

**Comment by aadhar-agarwal** [comment]
@jcastill Could I get another review on this when you get some time? Thank you!

**Comment by jcastill** [comment]
@aadhar-agarwal done! But you need another maintaner to add their review, mine is worth nothing.
@pmoravec @arif-ali can you check this when you have time?

**Comment by arif-ali** [comment]
I'm curious, the AzurePlugin has a lot of similar items for collection to RedHatPlugin, would it be worth extending the `AzureKdump` on top of `RedhatKdump`?

**Comment by pmoravec** [comment]
> I'm curious, the AzurePlugin has a lot of similar items for collection to RedHatPlugin, would it be worth extending the `AzureKdump` on top of `RedhatKdump`?

That makes sense to me. Or maybe even have one common plugin for both distros?

The diffs are
- `get_vm_core` option in Azure (can be beneficial for RedHat as well, I guess)
- RedHat parses more config files from `/etc/kdump.conf` (does it harm for Azure..? I doubt)
- `lsinitrd` is conditionally collected in RedHat
- a few copy_specs differ (union of them does not harm)

Either inheritance or common plugin sounds good to me. Current approach "a lot of code is shared, some bits differ" is not ideal but acceptable to me.

**Comment by jcastill** [comment]
I thought the same the key here is that they are not using the fs mount option, so their_kdump_conffile is much simpler than the rh one, and I'm not sure if we could/should make that function inside the class check what distribution is running and return different vmcore targets depending on that.
That said, I need to review the RedHat side of this plugin because there may be some things we can improve post RHEL 9 - so we can have this PR merged and have Azure collect their kdump info now, and then I can take a look and see if Azure and RH ones can be merged into one class, if that's ok with you all.

**Comment by arif-ali** [comment]
> I thought the same the key here is that they are not using the fs mount option, so their_kdump_conffile is much simpler than the rh one, and I'm not sure if we could/should make that function inside the class check what distribution is running and return different vmcore targets depending on that. That said, I need to review the RedHat side of this plugin because there may be some things we can improve post RHEL 9 - so we can have this PR merged and have Azure collect their kdump info now, and then I can take a look and see if Azure and RH ones can be merged into one class, if that's ok with you all.

makes good sense, let's create a new issue out of this, so that it is not forgotten; then we can circle back, and make improvements

**Review by arif-ali — APPROVED** [review]

**Pull request merged**
